// sweet-collections/dist/map/LfuMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LfuMap = void 0;
const SizedMap_1 = require("./SizedMap");
class LfuMap extends SizedMap_1.SizedMap {
    constructor(limit) {
        super(limit);
        this.clear();
    }
    has(key) {
        const node = this.map.get(key);
        if (node) {
            this.removeNode(node);
            this.moveToTop(node);
            return true;
        }
        return false;
    }
    get(key, _default) {
        const node = this.map.get(key);
        if (node) {
            this.removeNode(node);
            this.moveToTop(node);
            return node.value;
        }
        return _default;
    }
    set(key, value) {
        let node = this.map.get(key);
        if (node) {
            node.value = value;
            this.removeNode(node);
            this.moveToTop(node);
        }
        else {
            node = {
                key: key,
                value: value,
                used: 0,
            };
            if (this.isFull()) {
                super.delete(this.queuePop().key);
            }
            this.moveToTop(node);
            this.map.set(key, node);
        }
        return this;
    }
    delete(key) {
        const node = this.map.get(key);
        if (node) {
            this.removeNode(node);
            return super.delete(key);
        }
        return false;
    }
    clear() {
        this.queue = {};
        this.frequencies = { 1: { index: 1, count: 0 } };
        this.leastFrequent = this.frequencies[1];
        super.clear();
    }
    initQueue(used) {
        if (!this.queue[used]) {
            this.queue[used] = {};
        }
    }
    queuePop() {
        const min = this.leastFrequent.index;
        const node = this.queue[min].end;
        this.removeNode(node);
        return node;
    }
    moveToTop(node) {
        const used = ++node.used;
        this.increaseFrequency(used);
        this.initQueue(used);
        node.right = this.queue[used].start;
        node.left = undefined;
        if (this.queue[used].start) {
            this.queue[used].start.left = node;
        }
        this.queue[used].start = node;
        if (!this.queue[used].end) {
            this.queue[used].end = this.queue[used].start;
        }
    }
    removeNode(node) {
        const used = node.used;
        this.decreaseFrequency(used);
        this.initQueue(used);
        if (node.left) {
            node.left.right = node.right;
        }
        else {
            this.queue[used].start = node.right;
        }
        if (node.right) {
            node.right.left = node.left;
        }
        else {
            this.queue[used].end = node.left;
        }
    }
    increaseFrequency(index) {
        if (!this.frequencies[index]) {
            this.frequencies[index] = { index: index, count: 0 };
            this.frequencies[index].right = this.frequencies[index - 1].right;
            this.frequencies[index].left = this.frequencies[index - 1].count
                ? this.frequencies[index - 1]
                : this.frequencies[index - 1].left;
        }
        if (!this.frequencies[index].count) {
            if (this.frequencies[index].left) {
                this.frequencies[index].right = this.frequencies[index].left.right;
                this.frequencies[index].left.right = this.frequencies[index];
            }
            if (this.frequencies[index].right) {
                this.frequencies[index].left = this.frequencies[index].right.left;
                this.frequencies[index].right.left = this.frequencies[index];
            }
        }
        this.frequencies[index].count++;
        if (!this.leastFrequent || index < this.leastFrequent.index) {
            this.leastFrequent = this.frequencies[index];
        }
    }
    decreaseFrequency(index) {
        this.frequencies[index].count--;
        if (!this.frequencies[index].count) {
            if (this.frequencies[index].left) {
                this.frequencies[index].left.right = this.frequencies[index].right;
            }
            if (this.frequencies[index].right) {
                this.frequencies[index].right.left = this.frequencies[index].left;
            }
        }
        if (!this.leastFrequent || !this.leastFrequent.count) {
            this.leastFrequent = this.leastFrequent.right;
        }
    }
}
exports.LfuMap = LfuMap;
